src/hotspot/share/opto/mempointer.cpp

```diff
@@ -186,9 +216,70 @@ void MemPointerDecomposedFormParser::parse_sub_expression(const MemPointerSumman
   _summands.push(summand);
 }
 
+bool MemPointerParser::sub_expression_has_native_base_candidate(Node* start) {
+  // BFS over the expression.
+  // Allocate sufficient space in worklist for 100 limit below.
+  ResourceMark rm;
+  GrowableArray<Node*> worklist(102);
+  worklist.append(start);
+  for (int i = 0; i < worklist.length(); i++) {
+    Node* n = worklist.at(i);
+    switch(n->Opcode()) {
+      case Op_AddL:
+        // Traverse to both inputs.
+        worklist.append(n->in(1));
+        worklist.append(n->in(2));
+        break;
+      case Op_SubL:
+      case Op_CastLL:
+        // Traverse to the first input. The base cannot be on the rhs of a sub.
+        worklist.append(n->in(1));
+        break;
+      default:
+        if (is_native_memory_base_candidate(n)) { return true; }
+        break;
+    }
+    // This is a heuristic, so we are allowed to bail out early if the graph
+    // is too deep. The constant is chosen arbitrarily, not too large but big
+    // enough for all normal cases.
+    if (worklist.length() > 100) { return false; }
+  }
+  // Parsed over the whole expression, nothing found.
+  return false;
+}
+
+// Check if the node is a candidate to be a memory segment "base".
+// (1) CastX2P: some arbitrary long that is cast to a pointer.
+// (2) LoadL from field jdk.internal.foreign.NativeMemorySegmentImpl.min
+//     Holds the address() of a native memory segment.
+bool MemPointerParser::is_native_memory_base_candidate(Node* n) {
+  // (1) CastX2P
+  if (n->Opcode() == Op_CastX2P) { return true; }
+
+  // (2) LoadL from field jdk.internal.foreign.NativeMemorySegmentImpl.min
+  if (n->Opcode() != Op_LoadL) { return false; }
+  LoadNode* load = n->as_Load();
+
+  const TypeInstPtr* inst_ptr = load->adr_type()->isa_instptr();
+  if (inst_ptr == nullptr) { return false; }
+
+  ciInstanceKlass* klass = inst_ptr->instance_klass();
+  int offset = inst_ptr->offset();
+  ciField* field = klass->get_field_by_offset(offset, false);
+  if (field == nullptr) { return false; }
+
+  Symbol* field_symbol = field->name()->get_symbol();
+  Symbol* holder_symbol = field->holder()->name()->get_symbol();
+  if (holder_symbol != vmSymbols::jdk_internal_foreign_NativeMemorySegmentImpl() ||
+      field_symbol != vmSymbols::min_name()) {
+    return false;
+  }
+  return true;
```

Comment:
Could be simplified to:
```suggestion
  return holder_symbol == vmSymbols::jdk_internal_foreign_NativeMemorySegmentImpl() &&
         field_symbol == vmSymbols::min_name();
```